extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java

```diff
@@ -48,4 +87,77 @@ public String getImplementationName() {
     public ParquetInstructions readInstructions() {
         return readInstructions;
     }
+
+    /**
+     * Precedence-wise this implementation compares {@code order}, then applies a {@link PartitionsComparator} to
+     * {@code partitions}, then {@code dataSequenceNumber}, then {@code fileSequenceNumber}, then
+     * {@code manifestSequenceNumber}, then {@code dataFilePos}, and finally compares {@code uri}.
+     *
+     * @inheritDoc
+     */
+    @Override
+    public int compareTo(@NotNull final TableLocationKey other) {
+        if (other instanceof IcebergTableParquetLocationKey) {
+            // TODO(deephaven-core#5989): Add unit tests for the ordering of data files
+            final IcebergTableParquetLocationKey otherTyped = (IcebergTableParquetLocationKey) other;
+            int comparisonResult;
+            if ((comparisonResult = Integer.compare(order, otherTyped.order)) != 0) {
+                return comparisonResult;
+            }
+            if ((comparisonResult = PartitionsComparator.INSTANCE.compare(partitions, otherTyped.partitions)) != 0) {
+                return comparisonResult;
+            }
+            if (dataSequenceNumber != null && otherTyped.dataSequenceNumber != null &&
+                    (comparisonResult = Long.compare(dataSequenceNumber, otherTyped.dataSequenceNumber)) != 0) {
+                return comparisonResult;
+            }
+            if (fileSequenceNumber != null && otherTyped.fileSequenceNumber != null &&
+                    (comparisonResult = Long.compare(fileSequenceNumber, otherTyped.fileSequenceNumber)) != 0) {
+                return comparisonResult;
+            }
+            if ((comparisonResult = Long.compare(manifestSequenceNumber, otherTyped.manifestSequenceNumber)) != 0) {
+                return comparisonResult;
+            }
+            if ((comparisonResult = Long.compare(dataFilePos, otherTyped.dataFilePos)) != 0) {
+                return comparisonResult;
+            }
+            return uri.compareTo(otherTyped.uri);
+        }
+        throw new ClassCastException("Cannot compare " + getClass() + " to " + other.getClass());
+    }
+
+    @Override
+    public boolean equals(@Nullable final Object other) {
+        if (this == other) {
+            return true;
+        }
+        if (!(other instanceof IcebergTableParquetLocationKey)) {
+            return false;
+        }
+        // Iceberg devs have confirmed that uri's are supposed to be unique across data files
+        // https://apache-iceberg.slack.com/archives/C03LG1D563F/p1731352244907559
+        final IcebergTableParquetLocationKey otherTyped = (IcebergTableParquetLocationKey) other;
+        return uri.equals(otherTyped.uri);
+    }
```

Comment:
Even though we the devs are saying uri is guaranteed to be unique, we've seen that we may not be able to rely on that. As such, I think we need to be more defensive at this lower layer. Higher layers can choose to disallow duplicated URIs, but I don't think as the key we can.